pkg/workflow/controllers/task_reconciler.go

```diff
@@ -120,7 +120,11 @@ func (it *TaskReconciler) Reconcile(ctx context.Context, request reconcile.Reque
 
 	// update the status about conditional tasks
 	if len(pods) > 0 && (pods[0].Status.Phase == corev1.PodFailed || pods[0].Status.Phase == corev1.PodSucceeded) {
-		if !conditionalBranchesEvaluated(node) {
+		evaluated, err := it.conditionalBranchesEvaluated(ctx, node)
```

Comment:
> I looked at the new `conditionalBranchesEvaluated` function and it looks like the part added is a duplicate of the line above?

Yes. The condition is quite similar, but the usage is different. I have no good idea about how to refactor them in this PR.

Explain on logic, the first condition expression `len(pods) > 0 && (pods[0].Status.Phase == corev1.PodFailed || pods[0].Status.Phase == corev1.PodSucceeded) ` is used for determine the Pod existed and completed(succeed of failed), then invoke the next step: call the collector. The second expressions determine that evaluation has been finished(common routine: Pod Completed -> collectors collect the data -> evaluation the conditional branches), so the precondition of "evaluation finished" is "pod completed".

Explain on the code structure, the second expression contains in method `conditionalBranchesEvaluated()` which would be invoked **out of the scope of the first expression**. So only in a few situations, they are overlapped.

Actually, I just have an idea about splitting the several steps of `TaskNode` into different "reconcilers", like what "common", "condition", "desiredphase" and other chaos reconcilers do. We could do that, but I'm not 100% sure about the advantages of that, I prefer to use one “complete” huge reconcile rather than several reconcilers for now.

Reply:
I'll look at it right away.

Reply:
Great explanation! Now I understand the overall logic better than before. LGTM.